[IV-23-24] Objetivo 4

raultl12 opened the pull request:
* [x]  ¿Hay un camino claro que permita ir desde el código hasta la historia de
   usuario que desarrolla, pasando por el mensaje de commit y el issue
   correspondiente?
* [x]  ¿Se ha intentado cubrir con tests una parte de la lógica de negocio que se
   esté desarrollando?
* [x]  ¿Hay un producto mínimamente viable que describa lo que se va a entregar en
   este objetivo?
* [x]  ¿En este producto mínimamente viable se han priorizado unas clases,
   módulos o paquetes más fundamentales frente a otros, movidos a otro
   PMV/Hito?
* [ ] ¿Estoy leyendo todo lo que marco o simplemente marco lo que se me pone por
delante?
* [x] ¿Se han documentado las elecciones de biblioteca de aserciones y del
   de *test runner*?
* [x] ¿Te has asegurado que lo que mencionas como biblioteca de aserciones y
test runners lo son realmente y tienen la misma funcionalidad? ¿O estás poniendo
`unittest` como *test runner* y `pytest` como biblioteca de aserciones?
* [x] ¿Esa documentación incluye criterios de aceptación y también criterios de
búsqueda para las diferentes opciones?
* [x]  ¿Se han seguido los principios F.I.R.S.T. en la creación del test y
   se ha documentado cómo se ha hecho?
* [x]  ¿Se describe claramente en el PMV los tests que hay que pasar para que se
   considere viable?
* [x] ¿Se han testeado también las excepciones?

raultl12 committed:
doc: elegida herramienta para tests closes #35

Se ha elegido la herramienta que se usará para desarrollar y ejecutar tests que comprueben el funcionamiento del código.

raultl12 committed:
feat: importada biblioteca de aserciones closes #37

Se ha añadido a deno.json el import de la biblioteca de aserciones usada para hacer los tests

raultl12 committed:
feat: creado archivo types para separar la declaracion de enums closes #41

La declaracion de enumerados del archivo usuario se han movido al archivo types, añadiendo al archivo de usuario los import correspondientes

raultl12 committed:
feat: definidos entrenamientos closes #40

Se ha creado la clase entrenamiento para representar los entreamientos que serán asignados a los usuarios en funcion de los parámetros que establezcla la logica de negocio

raultl12 committed:
refactor: cambiada accesibilidad de campos de usuario closes #42

Se ha cambiado la accesibilidad de los campos de la clase usuario de publica a privada. Que los campos sean públicos sin un motivo de peso no es una buena práctica.

raultl12 committed:
feat: añadidos getters a usuario y entrenamiento closes #46

Se han añadido getters a las clases usuario y entrenamiento para que la lógica de negocio pueda leer estos atributos

raultl12 committed:
feat: añadidos contructores a usuario y entrenamiento closes #47

Se han añadido los constructores a las cases usuario y entrenamiento para que se puedan instanciar objetos

raultl12 committed:
feat: añadida clave test a iv.yaml closes #38

Se ha añadido al fichero iv.yaml la clave test indicando el fichero donde se encuentran los test sobre la logica de negocio

raultl12 committed:
feat: añadida tarea para ejecutar test closes #39

Se ha añadido al archivo de velociraptor la tarea para ejecutar los tests

raultl12 committed:
feat: creado asignador closes #49

Se ha creado la clase asignador donde se desarrollará lo lógica de negocio para asignar un entrenamiento a un usuario

raultl12 committed:
feat: añadida lógica de asignación closes #50

Se ha añadido al asignador la lógica necesaria para asignar el entrenamiento más adecuado a un usuario

raultl12 committed:
feat: creados los tests para el asignador closes #51

Se han creado los tests que comprueban el funcionamiento del asignador

raultl12 committed:
doc: actualizado readme con info sobre tests closes #52

Se ha modificado el readme añadiendo informacion sobre los tests ademas de el comando para ejecutarlos

raultl12 committed:
doc: modificado apartado clase closes #53

Se ha cambiado el apartado donde se indica la clase principal, no es usuario, sino asignador

raultl12 committed:
refactor: cambiada tarea check closes #54

Se ha cambiado la tarea check en velociraptor para que compruebe la sintaxis de la clase asignador y no la de usuario

adelahera reviewed (commented):
Hola, soy uno de tus revisores. Leyendo tu historia de usuario entiendo que el objetivo es mejorar el impacto de los jugadores. Hasta lo que he visto de momento, lo único que hace tu lógica de negocio es asignar un entrenamiento al jugador que le conviene. También entiendo que le prometes que con estos entrenamientos mejorará su impacto en el juego, pero quizás deberías explicar mejor eso, pues en la HU no se habla en ningún momento de asignar entrenamientos, simplemente que quiere mejorar su impacto.

adelahera commented on src/asignador.ts:
```diff
@@ -0,0 +1,82 @@
+        this.entrenamientos.push(entrenamiento);
+    }
+
+    asignarEntrenamiento(): Entrenamiento{
```
También te recomiendo modularizar un poco mejor el código. En vez de hacerlo todo en una misma función, es una buena práctica crear funciones específicas y luego llamarlas en esta función. Ayuda mucho a la hora de entender que estás haciendo en la función, además de que se te quedaría una función mucho más limpia. También, si en un futuro quieres ampliar la lógica de negocio y te hace falta implementar algo de lo que has hecho aquí, puedes hacerlo llamando a esa función, lo cual evitaría repetir código.

raultl12 replied:
Resuelto en: https://github.com/raultl12/TeamFinder/pull/36/commits/73c145f34dac77277f7858c1327f56766d8deaa3

adelahera commented on test/test_asignador.ts:
```diff
@@ -0,0 +1,67 @@
+  asignador.aniadirEntrenamiento(entrenamiento6);
+  asignador.aniadirEntrenamiento(entrenamiento7);
+
+  assertEquals(asignador.asignarEntrenamiento(), entrenamiento1);
```
Aquí en los tests, veo que simplemente estás comprobando que tu función devuelve el entrenamiento que debería. Creo que también debes comprobar que tu lógica de negocio supone una mejora frente a cualquier otra lógica. Por ejemplo, asigna un entrenamiento aleatorio y demuestra que tu lógica de negocio asigna el entrenamiento correcto, y la otra no.

raultl12 replied:
Resuelto en: https://github.com/raultl12/TeamFinder/pull/36/commits/ed3e1bc5a8c400424a4570d64a1337f20ff98e61

adelahera commented on test/test_asignador.ts:
```diff
@@ -0,0 +1,67 @@
+  asignador.aniadirEntrenamiento(entrenamiento6);
+  asignador.aniadirEntrenamiento(entrenamiento7);
+
+  assertEquals(asignador.asignarEntrenamiento(), entrenamiento3);
```
Aquí igual, lo mismo que te acabo de comentar.

raultl12 replied:
Resuelto en: https://github.com/raultl12/TeamFinder/pull/36/commits/ed3e1bc5a8c400424a4570d64a1337f20ff98e61

raultl12 committed:
refactor: modularizacion de asignador closes #55

Se han creado funciones para modularizar el codigo de asignación y que sea mas legible

raultl12 committed:
doc: actualizada HU closes #56

Se ha actualizado la HU indicando que Mario quiere que le se asignen entrenamientos personalizados.

raultl12 committed:
feat: añadido test para entrenamiento aleatorio closes #57

Se ha añadido un nuevo test que comprueba que el entrenamiento que devuelve la lógica de negocio es distinto a un entrenamiento seleccionado al azar.

raultl12 commented:
@JJ listo para revision

JJ commented:
@adelahera ¿has aprobado esto?

adelahera commented:
> @adelahera ¿has aprobado esto?

Pues no, se me había pasado, no se por qué pero no me llegó la notificación. De todas formas, le acabo de echar un vistazo y creo que está bastante bien.

JJ commented on docs/test.md:
```diff
@@ -0,0 +1,19 @@
+## Herramientas disponibles
+* Deno testing: Esta herramienta ya viene instalada con el runtime del proyecto, por lo que no se deberia instalar ninguna biblioteca externa, además, cuenta con biblioteca de aserciones propia, por lo que no hay que instalarla, solo importarla. 
+
+* Oak: Es especifico para Deno, tampoco requiere de instalación pero es más complejo que deno test pues requiere crear MockContext y MockAps. ([ver documentación](https://deno.land/x/oak@v12.6.1/docs/testing.md?source=))
```
Es un framework de middleware para servicios web. Lo que has enlazado son los documentos que sirven para testearlo, creo.

raultl12 replied:
Resuelto en: https://github.com/raultl12/TeamFinder/pull/36/commits/5a36d732a015780aa38272f82c68787fd2482ffb

JJ reviewed (changes requested):
Es esencial en este objetivo que las herramientas que examinéis se correspondan efectivamente al concepto que se está buscando. Por favor, hazlo así. Y enlaza todas las herramientas de las que hayas hablado.

raultl12 committed:
doc: eliminado oak como herramienta para test closes #58

Se ha eliminado oak como herramienta para test porque es un middleware de deno para http. Se habia confundido el testing del middleware con un framework completo

raultl12 committed:
doc: eliminado jest de test closes #60

Se ha eliminado jest porque requiere de npm para su instalación y ejecución.

raultl12 committed:
doc: añadido mocha al archivo de test closes #59

Se ha añadido mocha al archivo de test ya que no estaba contemplado como alternativa de framework para testing

raultl12 committed:
doc: añadida documentacion en archivo de test closes #61

Se han añadido enlaces de documentación e información sobre las herramientas disponibles

raultl12 commented:
@JJ listo para revisión

JJ commented on docs/test.md:
```diff
@@ -0,0 +1,17 @@
+* Deno testing: Esta herramienta ya viene instalada con el runtime del proyecto, por lo que no se deberia instalar ninguna biblioteca externa, además, cuenta con biblioteca de aserciones propia, por lo que no hay que instalarla, solo importarla. Ver [documentación](https://docs.deno.com/runtime/manual/basics/testing/)
+
+* Mocha: Es un framework de testing para node.js pero que podemos usar con deno. No es necesario instalarlo solo importarlo. No dispone de biblioteca de aserciones, pero funciona muy bien con [Chai](https://www.chaijs.com/). El problema de Chai es que requiere de npm, por lo que podemos usar la biblioteca de aserciones de deno.Ver [Deno mocha](https://deno.land/x/deno_mocha@0.3.1) y [aserciones e ...
+
```
Hay otras herramientas disponibles, incluso con versiones para deno. Por favor, dale una pequeña vuelta más.

raultl12 replied:
Resuelto en: https://github.com/raultl12/TeamFinder/pull/36/commits/bf36dd5c59175cc499dc20400de1be37b8880084

JJ commented on src/asignador.ts:
```diff
@@ -0,0 +1,103 @@
+    private filtroHabilidad: TipoHabilidad;
+
+    constructor(){
+        this.usuario = null;
```
Es una buena práctica asegurarte que los constructores nunca devuelvan objetos en un estado incorrecto.

raultl12 replied:
Resuelto en: https://github.com/raultl12/TeamFinder/pull/36/commits/07e3e4812085193258b57517679522ed45b97267

JJ reviewed (changes requested):
Iba a continuar, pero ya he detectado algunos problemas con el código. Mira [los errores frecuentes](https://github.com/JJ/IV-/blob/master/errores/errores-objetivo-4.md#sobre-la-programaci%C3%B3n), especialmente en el apartado de programación. También en los del [objetivo 2](https://github.com/JJ/IV-/blob/master/errores/errores-objetivo-2.md#sobre-el-c%C3%B3digo-definido) lo que dice sobre los setters:

> Sólo hay que incluir en el código lo que sea necesario para entender el modelo. Los setters, y mucho menos los getters, en general no van a ser necesarios.

Y no lo serán porque la construcción de un objeto se debe hacer en el constructor, que debe devolver un objeto construido o fallar.

Por otro lado, aunque no he llegado ahí, [comprueba que el error principal sobre los tests no ocurre aquí](https://github.com/JJ/IV-/blob/master/errores/errores-objetivo-4.md#sobre-los-tests)

raultl12 committed:
doc: añadidas nuevas herramientas de testing closes #62

Se han añadido nuevas herramientas de testing para completar el archivo de test.md

raultl12 committed:
refactor: corregido constructor de asignador closes #63

Se ha corregido el constructor del asignador para que siempre devuelva objetos válidos

raultl12 committed:
refactor: eliminados setters en asignador closes #64

Se han eliminado los setters innecesarios del asignador porque habían quedado inválidos por la mejora del constructor

raultl12 committed:
refactor: Corregidos test closes #65

Se han corregido los test tras el cambio del constructor del asignador

raultl12 committed:
feat: Mejorados test closes #66

Se han mejorado los test haciendo que comprueben si el entrenamiento devuelto por la logica de negocio esta entre los entrenamientos que mas cualidades tienen en comun con el usuario. Si esto se cumple, el resultado devuelto por la logica es valido

raultl12 commented:
@JJ Listo para revision

raultl12 committed:
refactor: buenas practicas en declaracion de variables closes #67

Se ha cambiado la declaracion de las variables para que use las buenas practicas

raultl12 committed:
refactor: eliminado setter no usado closes #69

Se ha eliminado el setter del filtro de habilidad ya que no estaba siendo usado

raultl12 committed:
refactor: buenas practicas en getters closes #68

Se han cambiado los getter de asignador, entrenamiento y usuario para que sigan las buenas practicas.

raultl12 commented:
@JJ Listo para revision

JJ reviewed (approved):
👍

raultl12 committed:
refactor: Corregido test aleatorio closes #70

Se ha corregido el test que asignaba un entrenamiento de forma aleatoria para demostrar que ese entrenamiento no es el adecuado
